### DISCUSSION TITLE
Alternating sign in eigenvector

### DISCUSSION THREAD

**Discussion by HelgeGehring** [kinnala/scikit-fem#970] [Q&A]
I'm trying to implement the equation
![image](https://user-images.githubusercontent.com/42973196/201446614-36e9b85f-fba1-4cf0-a3ac-c8c46d566869.png)
as follows:

```python
import scipy.linalg

from skfem import *
from skfem.helpers import *

mesh = MeshLine(np.linspace(-5, 5, 201))
mesh = mesh.with_subdomains({'core': lambda p: abs(p[0]) < .22})

basis_epsilon = Basis(mesh, ElementLineP0())

epsilon = basis_epsilon.zeros() + 1.444 ** 2
epsilon[basis_epsilon.get_dofs(elements='core')] = 3.4777 ** 2
basis_epsilon.plot(epsilon).show()

wavelength = 1.55
k0 = 2 * np.pi / wavelength


@BilinearForm
def lhs(u, v, w):
    return 1 / k0 ** 2 * dot(grad(u), grad(v)) + w['epsilon'] * inner(u, v)


@BilinearForm
def rhs(u, v, w):
    return inner(u, v)


basis = basis_epsilon.with_element(ElementLineP1())

A = lhs.assemble(basis, epsilon=basis_epsilon.interpolate(epsilon))
B = rhs.assemble(basis)

lams, xs = scipy.linalg.eigh(A.todense(), B.todense())
# lams, xs = solve(*condense(A, B, D=basis.get_dofs()), solver=solver_eigen_scipy(sigma=3.55 ** 2, which='LR'))

print(lams)
basis.plot(xs[:, -1]).show()
basis.plot(np.abs(xs[:, -1])).show()
```

The result when I plot the mode is
![image](https://user-images.githubusercontent.com/42973196/201446513-8e0b6be5-8353-4c85-9d2a-8da3569027d7.png)
as it oscillates between positive and negative.
If I plot only the real part I get
![image](https://user-images.githubusercontent.com/42973196/201446554-99965cfa-105d-48d2-ba4f-312cf1a1f838.png)
which seems totally fine to me.

Besides that, I'd also expect all eigenvalues to be smaller than `3.4777 ** 2`, but most of them are way larger.
Would you have an idea what is going wrong?

**Comment by kinnala** [comment]
Edit: I changed this answer because I originally misread your code.

Now I don't see what's wrong with the result. You are drawing the mode corresponding to the largest eigenvalue and AFAIK it should be highly oscillating. If you want a less oscillating mode, just draw the first mode:
```python
basis.plot(xs[:, 0]).show()
```

**Reply by kinnala** [reply]
Are you sure about the equation? I think when you integrate by parts the sign will change.

**Reply by kinnala** [reply]
I actually did change your basis definitions to
```python
basis = Basis(mesh, ElementLineP1())
basis_epsilon = basis.with_element(ElementLineP0())
```
It might've not changed the result. However, if you do it otherway around you might end up with less quadrature points that is required.

**Reply by HelgeGehring** [reply]
Thanks! Somehow I must have lost that sign while testing, thanks! Now it works fine :)

About the quadrature points:
```python
print(Basis(mesh, ElementLineP0()).quadrature)
print(Basis(mesh, ElementLineP1()).quadrature)
```
results in
```
(array([[0.21132487, 0.78867513]]), array([0.5, 0.5]))
(array([[0.21132487, 0.78867513]]), array([0.5, 0.5]))
```
Therefore, I guessed, that it doesn't matter with which element I define the mesh. Or do I miss here something?

Or is it more like a general rule: Define the first basis with the most complex Element and then define starting with that the other Basises?

**Reply by kinnala** [reply]
Alright, this must be a special case. In general, the quadrature order is calculated from basis function degree and Basis.with_element reuses the quadrature rule from the previous basis so that they are compatible inside forms.

**Reply by HelgeGehring** [reply]
Might it make sense to have a warning if I use `with_element` with a basis with higher function degree?
And why does the constant element have two quadrature points? The values are anyway the same, right?

**Reply by kinnala** [reply]
Here is the line: https://github.com/kinnala/scikit-fem/blob/f38541357605e6e19e6915341a8f6e58eeb42864/skfem/quadrature.py#L2825

Probably because otherwise linear form like (f, v) would be also integrated with just one point. We cannot automatically detect the degree of user provided data f so there is better to be more points than necessary just in case.

**Reply by HelgeGehring** [reply]
Okay, thanks :)
